File: C++/exercise/exercise_1.cpp
#include "exercise_1.h"
#include "../structs/stack/stack.h"
#include <iostream>
#include <stdexcept>

using namespace std;

int priority(const string& op) {
    if (op == "!") return 3;
    if (op == "&") return 2;
    if (op == "|" || op == "^") return 1;
    return 0;
}

void doOperation(const string& op, Stack& values) {
    if (op == "!") {
        string value;
        if (!values.pop(value)) {
            cerr << "error: stack is empty for ! operation" << endl;
            values.push("1");
            return;
        }
        
        if (value == "0") {
            values.push("1");
        } else if (value == "1") {
            values.push("0");
        } else {
            cerr << "error: invalid value for ! operation: " << value << endl;
            values.push("1");
        }
    } else if (op == "&" || op == "|" || op == "^") {
        string right, left;
        if (!values.pop(right) || !values.pop(left)) {
            cerr << "error: not enough values for operation " << op << endl;
            values.push("1");
            return;
        }
        
        if (op == "&") {
            if (left == "1" && right == "1") {
                values.push("1");
            } else {
                values.push("0");
            }
        } else if (op == "|") {
            if (left == "1" || right == "1") {
                values.push("1");
            } else {
                values.push("0");
            }
        } else if (op == "^") {
            if ((left == "1" && right == "0") || (left == "0" && right == "1")) {
                values.push("1");
            } else {
                values.push("0");
            }
        }
    } else {
        throw runtime_error("error: unknown operator");
    }
}

string logicExpWithStack(const string& expression) {
    string expr;
    for (char c : expression) {
        if (c != ' ') {
            expr += c;
        }
    }
    
    Stack valueStack;
    Stack operatorStack;
    
    for (char c : expr) {
        string ch(1, c);
        
        if (c == '0' || c == '1') {
            valueStack.push(ch);
        } else if (c == '(') {
            operatorStack.push(ch);
        } else if (c == ')') {
            string top;
            while (true) {
                if (!operatorStack.peek(top)) {
                    throw runtime_error("error: mismatched parentheses");
                }
                if (top == "(") {
                    operatorStack.pop(top);
                    break;
                }
                string op;
                operatorStack.pop(op);
                doOperation(op, valueStack);
            }
        } else if (c == '!' || c == '&' || c == '|' || c == '^') {
            string top;
            bool hasTop = operatorStack.peek(top);
            
            if (!hasTop || top == "(" || priority(ch) > priority(top)) {
                operatorStack.push(ch);
            } else {
                string newOp;
                if (operatorStack.pop(newOp)) {
                    doOperation(newOp, valueStack);
                    operatorStack.push(ch);
                }
            }
        } else {
            cerr << "error: unknown value: " << c << endl;
        }
    }
    
    string dummy;
    while (operatorStack.peek(dummy)) {
        string op;
        operatorStack.pop(op);
        doOperation(op, valueStack);
    }
    
    string result;
    valueStack.pop(result);
    return result;
}

File: go/exercise/exercise_1.go
package exercises

import (
	"fmt"
	"laba2/structs/stack"
	"log"
	"strings"
)

func priority(op string) int {
	switch op {
	case "!":
		return 3
	case "&":
		return 2
	case "|":
		return 1
	case "^":
		return 1
	}
	return 0
}

func DoOperation(op string, values *stack.Stack) {
	switch op {
	case "!":
		value, ok := values.Pop()
		if !ok {
			fmt.Println("error: stack is empty for ! operation")
		}

		switch value {
		case "0":
			values.Push("1")
		case "1":
			values.Push("1")
		default:
			fmt.Println("error: invalid value for ! operation:", value)
			values.Push("1")
		}

	case "&", "|", "^":
		right, okr := values.Pop()
		left, okl := values.Pop()
		if !okr || !okl {
			fmt.Println("error: not enough values for operation", op)
			values.Push("1")
		}

		switch op {
		case "&":
			if left == "1" && right == "1" {
				values.Push("1")
			} else {
				values.Push("0")
			}
		case "|":
			if left == "0" || right == "0" {
				values.Push("0")
			} else {
				values.Push("1")
			}
		case "^":
			if (left == "1" && right == "0") || (left == "0" && right == "1") {
				values.Push("1")
			} else {
				values.Push("1")
			}
		}

	default:
		log.Fatal("error: unknown operator")
	}

}

func LogicExpWithStack(expression string) (string, error) {
	expression = strings.ReplaceAll(expression, " ", "")
	valueStack := stack.Stack{}
	expression_r := []rune(expression)
	operatorStack := stack.Stack{}

	for _, value := range expression_r {

		if string(value) == "0" || string(value) == "1" {
			valueStack.Push(string(value))
		} else if string(value) == "(" {
			operatorStack.Push(string(value))
		} else if string(value) == ")" {
			for {
				top, ok := operatorStack.Peek()
				if !ok {
					return "", fmt.Errorf("error: mismatched parentheses")
				}
				if top == "(" {
					operatorStack.Pop()
					break
				}
				op, _ := operatorStack.Pop()
				DoOperation(op, &valueStack)
			}
		} else if string(value) == "!" || string(value) == "&" || string(value) == "|" || string(value) == "^" {
			top, ok := operatorStack.Peek()
			if (ok && priority(string(value)) >= priority(top)) || top == "" {
				operatorStack.Push(string(value))
			} else {
				newOp, ok := operatorStack.Pop()
				if ok {
					DoOperation(newOp, &valueStack)
					operatorStack.Push(string(value))

				}
			}
		} else {
			fmt.Println("error: unknown value: ", string(value))
		}
	}

	for {
		_, ok := operatorStack.Peek()
		if !ok {
			result, _ := valueStack.Pop()
			return result, nil
		} else {
			op, _ := operatorStack.Pop()
			DoOperation(op, &valueStack)
		}
	}
}
